### ISSUE TITLE
Proposal for X-Request-Id

### ISSUE THREAD

**Issue by DoumanAsh** [tower-rs/tower-http#135]
## Feature Request

Add hanlding of `x-request-id` and/or automatically generate it when missing

### Motivation

Request-id is useful if we want to be able to efficiently correlate logs between client and server.
If client doesn't generate himself, we still can benefit by sending such id from server as we can ask client later to provide it if there is something wrong with response

### Proposal

Implementation example https://github.com/DoumanAsh/tower-http-req-id

For now there is only UUID generator, but I made code generic enough to allow crate's user to supply own generators.

Basically idea is that middleware checks whether header is present, if it is then it stores it in extensions.
If not it generates and adds it to extension.

Following that once request is resolved it adds the same header to Response

### Alternatives

I'll publish it as 3pp crate

**Comment by DoumanAsh** [comment]
@davidbarsky ping, what do you think about adding such middleware to this crate?

**Comment by davidpdrsn** [comment]
I think it makes sense! I've actually played around with a request-id layer at work.

The API I have in mind is

```rust
struct SetRequestIdLayer {
    // ...
}

impl SetRequestId {
    fn uuid() -> Self;

    fn custom_generator<T>(gen: T) -> Self
    where
        T: IdGenerator;

    // Defaults to X-Request-Id
    fn header_name(self, name: HeaderName) -> Self;
}

trait IdGenerator {
    fn gen(&mut self) -> HeaderValue;
}
```

What do you think?

**Comment by DoumanAsh** [comment]
With API that you have in mind, how user access the id?
Should he always inspect header for when it is needed?
I don't really like extensions myself as they are quite ugly to use due to lack of compile-time awareness, but it makes your id typed at least, allowing a simple way to copy it when it is for example integer like uuid 😄

**Comment by davidpdrsn** [comment]
Hm good question. Certainly a `RequestId(pub HeaderValue)` extension would be easiest but certainly has drawbacks as you mention. Do you have any other ideas?

**Comment by DoumanAsh** [comment]
There is another way to make it less ugly for user:
1. Insert `X-Request-ID` if user client didn't set it.
2. Use extractor to get `id` from header
3. Profit

That should at least remove need for user to work directly with headers himself.
